File: go/p027.go
package main

import (
	"./lib/primes"
	"fmt"
)

const MAX = 1000

var prime_list = primes.List(MAX)

func main() {
	ma, mb, mc := 0, 0, 0
	for a := -999; a < 1000; a += 2 {
		for _, b := range prime_list {
			c := longest(a, b)
			if c > mc {
				ma, mb, mc = a, b, c
			}
		}
	}
	fmt.Println(ma*mb, ma, mb, mc)
	// output: -59231 -61 971 71
}

func longest(a, b int) int {
	for x := 0; x < b; x++ {
		f := x*x + a*x + b
		if !primes.Test(f) {
			return x
		}
	}
	return b
}

File: python/p027.py
# Quadratic primes
#
# Considering quadratics of the form:
#
#     n² + an + b, where |a| < 1000 and |b| < 1000
#
# Find the product of the coefficients, a and b, for the quadratic
# expression that produces the maximum number of primes for
# consecutive values of n, starting with n = 0.

# When n = 0, p(0) = b, hence b prime

import primes
p = primes.gen(1000)

def strike(a,b):
    for n in range(b):
        fn = n*n + a*n + b
        if not(primes.test(fn)):
            return n
    return b

ma, mb, mn = 0, 0, 0

for a in range(-999,1000,2):
    for b in p[12:]: # p[12] == 41, starting from there.
        n = strike(a,b)
        if n > mn:
            ma, mb, mn = a, b, n

print(ma*mb)
# output: -59231
